File: src/table_view.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include <pybind11/pybind11.h>
#include <pulsar/TableView.h>
#include <pulsar/Schema.h>
#include <pulsar/TableViewConfiguration.h>
#include <pybind11/stl.h>
#include <utility>
#include "utils.h"

namespace py = pybind11;
using namespace pulsar;

void export_table_view(py::module_& m) {
    py::class_<TableViewConfiguration>(m, "TableViewConfiguration")
        .def(py::init<>())
        .def("subscription_name",
             [](TableViewConfiguration& config, const std::string& name) { config.subscriptionName = name; })
        .def("schema",
             [](TableViewConfiguration& config, const SchemaInfo& schema) { config.schemaInfo = schema; });

    py::class_<TableView>(m, "TableView")
        .def(py::init<>())
        .def("get",
             [](const TableView& view, const std::string& key) -> std::pair<bool, py::bytes> {
                 py::gil_scoped_release release;
                 std::string value;
                 bool available = view.getValue(key, value);
                 py::gil_scoped_acquire acquire;
                 if (available) {
                     return std::make_pair(true, py::bytes(std::move(value)));
                 } else {
                     return std::make_pair(false, py::bytes());
                 }
             })
        .def("size", &TableView::size, py::call_guard<py::gil_scoped_release>())
        .def("close", [](TableView& view) {
            waitForAsyncResult([&view](ResultCallback callback) { view.closeAsync(callback); });
        });
}

File: pulsar/tableview.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

"""
The TableView implementation.
"""

from typing import Any, Callable, Optional
from pulsar.schema.schema import Schema
import _pulsar

class TableView():

    def __init__(self, table_view: _pulsar.TableView, topic: str,
                 subscription: Optional[str], schema: Schema) -> None:
        self._table_view = table_view
        self._topic = topic
        self._subscription = subscription
        self._schema = schema

    def get(self, key: str) -> Optional[Any]:
        """
        Return the value associated with the given key in the table view.

        Parameters
        ----------
        key: str
            The message key

        Returns
        -------
        Optional[Any]
            The value associated with the key, or None if the key does not exist.
        """
        pair = self._table_view.get(key)
        if pair[0]:
            return self._schema.decode(pair[1])
        else:
            return None

    def for_each(self, callback: Callable[[str, Any], None]) -> None:
        """
        Iterate over all entries in the table view and call the callback function
        with the key and value for each entry.

        Parameters
        ----------
        callback: Callable[[str, Any], None]
            The callback function to call for each entry.
        """
        self._table_view.for_each(lambda k, v: callback(k, self._schema.decode(v)))

    def for_each_and_listen(self, callback: Callable[[str, Any], None]) -> None:
        """
        Iterate over all entries in the table view and call the callback function
        with the key and value for each entry, then listen for changes. The callback
        will be called when a new entry is added or an existing entry is updated.

        Parameters
        ----------
        callback: Callable[[str, Any], None]
            The callback function to call for each entry.
        """
        self._table_view.for_each_and_listen(lambda k, v: callback(k, self._schema.decode(v)))

    def close(self) -> None:
        """
        Close the table view.
        """
        self._table_view.close()

    def __len__(self) -> int:
        """
        Return the number of entries in the table view.
        """
        return self._table_view.size()

    def __str__(self) -> str:
        if self._subscription is None:
            return f"TableView(topic={self._topic})"
        else:
            return f"TableView(topic={self._topic}, subscription={self._subscription})"

    def __repr__(self) -> str:
        return self.__str__()
